diff --git i/modules/extensions.py w/modules/extensions.py
index b522125..6ea02df 100644
--- i/modules/extensions.py
+++ w/modules/extensions.py
@@ -7,7 +7,7 @@ import git
 from modules import paths, shared
 
 extensions = []
-extensions_dir = os.path.join(paths.script_path, "extensions")
+extensions_dir = os.environ.get("SD_EXTENSIONS_DIR")
 extensions_builtin_dir = os.path.join(paths.script_path, "extensions-builtin")
 
 
