diff --git i/modules/generation_parameters_copypaste.py w/modules/generation_parameters_copypaste.py
index 46e12dc..5713e56 100644
--- i/modules/generation_parameters_copypaste.py
+++ w/modules/generation_parameters_copypaste.py
@@ -289,7 +289,7 @@ Steps: 20, Sampler: Euler a, CFG scale: 7, Seed: 965400086, Size: 512x512, Model
 def connect_paste(button, paste_fields, input_comp, jsfunc=None):
     def paste_func(prompt):
         if not prompt and not shared.cmd_opts.hide_ui_dir_config:
-            filename = os.path.join(script_path, "params.txt")
+            filename = os.environ.get("SD_PARAMS_FILE")
             if os.path.exists(filename):
                 with open(filename, "r", encoding="utf8") as file:
                     prompt = file.read()
diff --git i/modules/processing.py w/modules/processing.py
index bc541e2..06934e0 100644
--- i/modules/processing.py
+++ w/modules/processing.py
@@ -571,7 +571,7 @@ def process_images_inner(p: StableDiffusionProcessing) -> Processed:
             if not p.disable_extra_networks:
                 extra_networks.activate(p, extra_network_data)
 
-        with open(os.path.join(shared.script_path, "params.txt"), "w", encoding="utf8") as file:
+        with open(os.environ.get("SD_PARAMS_FILE"), "w", encoding="utf8") as file:
             processed = Processed(p, [], p.seed, "")
             file.write(processed.infotext(p, 0))
 
